compiler/optimizations/propagateDomainConstness.cpp

```diff
@@ -0,0 +1,249 @@
+    }
+  }
+}
+
+// Add code right after a PRIM_SET_MEMBER, if it is setting a const domain
+// member in an aggregate type
+void setDefinedConstForPrimSetMemberIfApplicable(CallExpr *call) {
+  INT_ASSERT(call->isPrimitive(PRIM_SET_MEMBER));
+
+  Expr *nextExpr = call->next;
+  bool addNoop = addNoopAsNextExprIfNeeded(call, nextExpr);
+
+  // PRIM_SET_MEMBER args:
+  // 1: this
+  // 2: field
+  // 3: src
+  Symbol *thisSym = NULL;
+  if (SymExpr *se = toSymExpr(call->get(1))) {
+    thisSym = se->symbol();
+  }
+
+  Symbol *fieldSym = NULL;
+  Type *fieldType = NULL;
+  if (SymExpr *se = toSymExpr(call->get(2))) {
+    fieldSym = se->symbol();
+    fieldType = fieldSym->getValType();
+  }
+  INT_ASSERT(fieldSym);
+  INT_ASSERT(fieldType);
+
+  if (!fieldSym->hasFlag(FLAG_TYPE_VARIABLE)) {
+    if (fieldType->symbol->hasFlag(FLAG_DOMAIN)) {
+      Symbol *isConst = fieldSym->hasFlag(FLAG_CONST) ? gTrue : gFalse;
+      setDefinedConstForDomainField(thisSym, fieldSym, nextExpr, isConst);
+    }
+  }
+
+  if (addNoop) {
+    nextExpr->remove();
+  }
+
+}
+
+// go through all PRIM_SET_MEMBER in an initializer, and adjust them if needed
+void setDefinedConstForFieldsInInitializer(FnSymbol *fn) {
+  std::vector<CallExpr *> calls;
+  collectCallExprs(fn->body, calls);
+  for_vector(CallExpr, call, calls) {
+    if(call->isPrimitive(PRIM_SET_MEMBER)) {
+      if (fn->_this == toSymExpr(call->get(1))->symbol()) {
+        setDefinedConstForPrimSetMemberIfApplicable(call);
+      }
+    }
+  }
+}
+
+// help removing chpl__initCopy and chpl__autoCopy after resolution while
+// retaining constness information therein
+void removeInitOrAutoCopyPostResolution(CallExpr *call) {
+  Expr *parentExpr = call->parentExpr;
+  Expr *nextExpr = parentExpr->next;
+  bool addNoop = addNoopAsNextExprIfNeeded(parentExpr, nextExpr);
+
+  Symbol *argSym = NULL;
+  Type *argType = NULL;
+  if (SymExpr *se = toSymExpr(call->get(1))) {
+    argSym = se->symbol();
+    argType = argSym->getValType();
+  }
+  INT_ASSERT(argSym);
+  INT_ASSERT(argType);
+
+  call->replace(call->get(1)->remove());
+
+  // we removed the first argument already, so definedConst is the first
+  // argument now
+  SymExpr *secondArg = toSymExpr(call->get(1)->remove());
+  INT_ASSERT(secondArg);
+
+  Symbol *isConst = secondArg->symbol();
+  INT_ASSERT(isConst->type == dtBool);
+
+  if (argType->symbol->hasFlag(FLAG_DOMAIN)) {
+    Symbol *lhs = NULL;
+    if (CallExpr *parentCall = toCallExpr(parentExpr)) {
+      if (parentCall->isPrimitive(PRIM_MOVE)) {
+        if (SymExpr *lhsSE = toSymExpr(parentCall->get(1))) {
+          lhs = lhsSE->symbol();
+        }
+      }
+    }
+    INT_ASSERT(lhs);
+
+    if (!isShadowVarSymbol(lhs)) {
+      Expr *anchor = nextExpr;
+      setDefinedConstForDomainSymbol(lhs, nextExpr, anchor, isConst);
+
+      if (addNoop) {
+        nextExpr->remove();
+      }
+    }
+  }
+}
+
+static bool addNoopAsNextExprIfNeeded(Expr *baseExpr, Expr *&nextExpr) {
```

Comment:
I think I would be less confused by this function if it always computed and returned `nextExpr`; sometimes setting it to a new noop. It could be called something like `getNextExprOrCreateNoop`. It would also make more sense to me if it returned the nextExpr and used a `bool & createdNoop`.